directory/directory_transport.go

```diff
@@ -70,11 +76,44 @@ type dirReference struct {
 // We do not expose an API supplying the resolvedPath; we could, but recomputing it
 // is generally cheap enough that we prefer being confident about the properties of resolvedPath.
 func NewReference(path string) (types.ImageReference, error) {
+	// In order to allow for having "compressed" and "uncompressed"
+	// directories, any path with a suffix of :: is treated as an option suffix
+	// (comma-separated).
+	var options []string
+	if strings.Contains(path, "::") {
+		parts := strings.SplitN(path, "::", 2)
+		if len(parts) != 2 {
+			// Coverage: should never happen
+			return nil, errors.Errorf("dir: invalid path option suffix: %s", path)
+		}
+		options = strings.Split(parts[1], ",")
+		if len(options) == 0 {
+			return nil, errors.Errorf("dir: invalid path option suffix: no options set: %s", path)
+		}
+		path = parts[0]
+	}
```

Comment:
If we did do so much parsing for `dir` references, this should probably move to the mode of `ParseReference` doing the string parsing, and `NewReference(path string, compressed bool)` just creating an object (so that API callers can do it cheaply).